[Diagnostics] Introduce argument-to-parameter mismatch

xedin opened the pull request:
This PR aims to port argument-to-parameter conversion failures in their most general form `cannot convert X to expected argument Y` and some port tailored diagnostics related to them, lift some of the restrictions on use of requirement failures and remove now obsolete logic from CSDiag.

Resolves: rdar://problem/54613698
Resolves: rdar://problem/55254652
Resolves: rdar://problem/51262694
Resolves: rdar://problem/52334290
Resolves: rdar://problem/17647738
Resolves: rdar://problem/18624434
Resolves: rdar://problem/19979645
Resolves: rdar://problem/21032147
Resolves: rdar://problem/22042549
Resolves: rdar://problem/25639952
Resolves: rdar://problem/28047208
Resolves: rdar://problem/28555798
Resolves: rdar://problem/28675189
Resolves: rdar://problem/29214149
Resolves: rdar://problem/31785308
Resolves: rdar://problem/32390841
Resolves: rdar://problem/33914446
Resolves: rdar://problem/41303412
Resolves: rdar://problem/35420209
Resolves: rdar://problem/21948825
Resolves: [SR-11388](https://bugs.swift.org/browse/SR-11388)
Resolves: [SR-4795](https://bugs.swift.org/browse/SR-4795)

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[CSFix] Introduce a fix for argument-to-parameter conversion mismatch

xedin committed:
[Diagnostics] Add generic argument-to-parameter conversion failure

No specialized diagnostics yet, just a basic infrastructure to
produce generic a `cannot convert X to expected argument type Y`
message.

xedin committed:
[Diagnostics] Add a tailored diagnostic for attempt to use optional as a `Bool` in argument position

xedin committed:
[Diagnostics] Detect situation when argument would match if it wasn't optional

xedin committed:
[ConstraintSystem] Score generic argument-to-parameter mismatch fixes down

Since this is the most generic failure, we need to make sure that
it doesn't interfere with other more specific fixes.

xedin committed:
[Diagnostics] Port a custom diagnostic for archetype mismatch in argument position

xedin committed:
[Diagnostics] Tailored diagnostic when argument/parameter mismatch is due to missing conformance

xedin committed:
[Diagnostics] Attempt fix-its appropriate for types in argument positions

Most of the fix-its associated with contextual failures are also
appropriate for argument-to-parameter conversion mismatches.

xedin committed:
[ConstraintSystem] Make sure that l-value is preserved in argument positions

There is logic in `matchTypes` which would unwrap l-value if other
type is not an `inout`, which is not great for cases where parameter
type is a pointer and argument is an l-value which requires explicit
`&` to be converted.

xedin committed:
[Diagnostics] Port pattern-matching mismatch diagnostic

Port diagnostics associated with implicit use of `~=` operator
which is used in `case` statements.

xedin committed:
[ConstraintSystem] Don't produce argument mismatch fix when pointer conversions are involved

xedin committed:
[ConstraintSystem] Fix a situation when argument requires raw representable type

xedin committed:
[ConstraintSystem] Fix a situation when argument requires raw value type of raw representable

xedin committed:
[CSDiag] NFC: Remove obsolete logic which deals with raw representable mismatches

xedin committed:
[ConstraintSystem] Rate impact of conformance failures based on non-conforming type use

Impact of conformance failure should be rated based on use of
associated generic parameter, because a single failure could
span multiple parameters and a result.

Originally score was only increased once for conformance failure.
That is not good enough in cases when conformance is associated
with an overload choice, because in that case we can have an argument
mismatch in other choice which would have the same score impact.

xedin committed:
[ConstraintSystem] Lift restriction which restricted conformance fixes in argument positions

If argument doesn't conform to the protocol expected by the parameter
let's add argument conversion mismatch fix for that with increased impact.

xedin committed:
[CSDiag] NFC: Remove obsolete `CalleeCandidateInfo::diagnoseGenericParameterErrors`

xedin committed:
[Diagnostics] Allow requirement failure to diagnose operators

In absence of general argument conversion failures requirement
errors associated with operators couldn't be diagnosed properly,
now this restriction could be lifted.

xedin committed:
[CSDiag] NFC: Remove obsolete `diagnoseArgumentGenericRequirements`

xedin committed:
[CSDiag] NFC: Remove more obosolete code from `visitApplyExpr`

xedin committed:
[Diagnostics] NFC: Update all of the improved test-cases

xedin commented:
This is still a draft, 10+ tests are failing, mainly due to changes related to `binary operator <op> cannot be applied to operands of type 'X' and 'Y'` diagnostic, which I need to port to `diagnoseAmbiguityWithFixes`.

xedin commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -379,6 +379,11 @@ ERROR(cannot_convert_default_arg_value_nil,none,
       (Type,Type))
+
+NOTE(candidate_has_invalid_argument_at_position,none,
+     "candidate expects value of type %0 at psoition #%1",
```
And of course there is a typo...

xedin committed:
[Diagnostics] Fix a typo in argument mismatch diagnostic note

slavapestov commented on lib/Sema/CSDiag.cpp:
```diff
@@ -3195,186 +3058,6 @@ bool FailureDiagnosis::diagnoseNilLiteralComparison(
-  if (candidates.closeness != CC_ExactMatch || candidates.size() != 1)
-    return false;
-
-  AbstractFunctionDecl *AFD = nullptr;
```
It looks like the old code here didn't handle generic subscripts at all, because it bails out if the candidate is not an AbstractFunctionDecl. Can you add some tests similar to the existing ones for functions, but for generic subscripts, and see if we produce better diagnostics now?

xedin replied:
Absolutely!

xedin replied:
I have added a couple of examples in bcd4180f302967155f78dc27708d1db8352561a3

slavapestov reviewed (commented):
The new diagnostics are looking much better overall. Excellent work!

xedin committed:
[ConstraintSystem] Fail comparison if dependent member types couldn't be resolved

Currently in cases where dependent member types cannot be resolved
properly due to invalid base type they do not fail comparison, but
instead result in a new "inactive" constraint which is incorrect.

xedin committed:
[Diagnostics] Don't diagnoses missing `Equatable` in pattern-matching of invalid expressions

If expression is incorrect it most likely wouldn't be able to satisfy
`Equatable` or other requirements of `~=` operator overloads, but
at the same time the main problem is related to `case` expression
itself so let's not diagnose missing conformances.

xedin committed:
[Diagnostics] Add special ambiguity diagnostics for operators

In cases when all of the fixed solutions have only one problem in
common - different overloads of a certain operator, let's
produce a tailored diagnostic and suggest matching partial
overloads along side diagnostic notes which point to each choice.

xedin committed:
[ConstraintSystem] Adjust a couple of places not to fail immediately in diagnostic mode

Some spots of constraint simplification logic are too eager to
fail right away without giving repair logic to run in diagnostic
mode and attempt to fix the problem. Failing early in diagnostic
mode means solver wouldn't be able to reach some possible
solutions which leads to subpar diagnostics.

slavapestov commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2928,10 +2928,17 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
+  // possible that base type is incorrect e.g. `U.Element` where `U`
+  // is a generic parameter, so checking equality here would lead to
+  // incorrect behavior, let's defer it until later proper check.
+  if (!(desugar1->is<DependentMemberType>() &&
```
Is the base type a generic parameter or concrete in this case? I'm a bit surprised to see DependentMemberTypes showing up here at all, since usually they only appear in interface types and constraint systems use contextual types?

xedin replied:
That's a consequence of `simplifyTypeImpl` implementation, it would re-create `DependentMemberType` if the base was incorrect concrete type. I think at some point we'd want to change that behavior but it could be done separately. It mostly shows up in diagnostics when missing conformances are allowed so for signatures like that `<T: Sequence, U: Sequence, T.Element == U.Element>` if we try `T = Int` and `U = String` for example, we'll get a requirement constraint as `Int.Element == String.Element` which used to be considered unsolved but now is going to fail.

slavapestov replied:
Oh, right. You might want to update the comment then -- the base will never be a generic parameter, but it might be a concrete type that does not conform to U.

I guess we have to allow this because simplifyType() cannot report failure, but I imagine if this happens then a conformance constraint somewhere else should have failed, correct?

xedin replied:
Yes, that's correct, it's never going to show up during normal solving, only in diagnostic mode and if conformance constraint fails.

xedin replied:
Sorry, comment should have been more clear about relationship between this problem and generic parameters.

xedin replied:
Updated comment to talk about concrete types instead.

xedin committed:
[Diagnostics] Provide a tailored diagnostic for operator mismatch on missing conformance

If missing conformance is between two stdlib defined types which
are used in operator invocation, let's produce a generic diagnostic
about operator reference and a note about missing conformance.

xedin committed:
[Diagnostics] Tailored diagnostics for reference equality operator mismatches

xedin committed:
[CSSimplify] NFC: Adjust comment about matching dependent types with invalid base

xedin committed:
[CSDiag] NFC: Remove obsolete fix-it for argument casts from `visitApplyExpr`

xedin committed:
[ConstraintSystem] Fix: Try to match argument to element type if parameter is a collection

This allows better diagnostics in cases where collection element
is a generic parameter which is supposed to be inferred from argument e.g.

```swift
func foo<T>(_: [T]) {}

foo(1) // Missing brackets, so error should be about [Int] vs. Int
```

xedin committed:
[Diagnostics] Fix usages of `getChoiceFor`

xedin committed:
[ConstraintSystem] De-prioritize missing member fix when base is contextual

Impact is higher if the base is expected to be inferred from context,
because a failure to find a member ultimately means that base type is
not a match in this case.

xedin committed:
[Diagnostics] NFC: Fix improved test-cases related to argument-to-parameter mismatches

xedin committed:
[Diagnostics] NFC: Add a couple of examples of requirement failures related to generic subscripts

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
Please test with following pull request:
https://github.com/apple/swift-package-manager/pull/2338

@swift-ci please test

xedin commented:
@shahmishal Doesn't look source compatibility suite failure in `CoreStore` is related to my changes:

```
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite-debug/project_cache/CoreStore/build/Build/Products/Debug-iphoneos/CoreStore.framework/Headers/CoreStore-Swift.h:1:2: error: unterminated conditional directive
#if 0
 ^
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite-debug/project_cache/CoreStore/build/Build/Products/Debug-iphoneos/CoreStore.framework/Headers/CoreStore-Swift.h:4151:37: error: missing '@end'
/// \param object the entity type fo
                                    ^
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite-debug/project_cache/CoreStore/build/Build/Products/Debug-iphoneos/CoreStore.framework/Headers/CoreStore-Swift.h:4138:1: note: protocol started here
@protocol CSListObjectObserver <CSListObserver>
^

/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite-debug/project_cache/CoreStore/Sources/CoreStoreBridge.m:182:13: error: receiver type 'CSSelect' for instance message is a forward declaration
    return [[CSSelect alloc] initWithDataTerm:selectTerm];
            ^~~~~~~~~~~~~~~~
In file included from /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite-debug/project_cache/CoreStore/Sources/CoreStoreBridge.m:27:
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite-debug/project_cache/CoreStore/Sources/CoreStoreBridge.h:362:8: note: forward declaration of class here
@class CSSelect;
       ^
```

DougGregor commented on lib/Sema/ConstraintSystem.h:
```diff
@@ -2063,7 +2063,7 @@ class ConstraintSystem {
   /// Log and record the application of the fix. Return true iff any
   /// subsequent solution would be worse than the best known solution.
-  bool recordFix(ConstraintFix *fix);
+  bool recordFix(ConstraintFix *fix, unsigned implact = 1);
```
Typo `implact` shows up a bit.

xedin replied:
Of course...

xedin replied:
Fixed!

DougGregor commented on test/Constraints/bridging.swift:
```diff
@@ -265,14 +261,12 @@ func rdar19831919() {
 // <rdar://problem/19831698> Incorrect 'as' fixits offered for invalid literal expressions
 func rdar19831698() {
-  var v70 = true + 1 // expected-error{{binary operator '+' cannot be applied to operands of type 'Bool' and 'Int'}} expected-note {{expected an argument list of type '(Int, Int)'}}
+  var v70 = true + 1 // expected-error{{cannot convert value of type 'Bool' to expected argument type 'Int'}}
```
With some of these cases, it might be worth splitting the line into separate lines so we can ensure that the diagnostic points at the correct operand, e.g., 

```
var v70 = true + 1 // expected-error{{cannot convert value of type 'Bool' to expected argument type 'Int'}}
```

could become

```
var v70 = true // expected-error{{cannot convert value of type 'Bool' to expected argument type 'Int'}}
  + 1 
```

xedin replied:
Good point!

owenv replied:
I don't remember if it's documented yet, but the verifier now supports specifying a column number via `expected-error@:42 {{...` which might come in handy here.

xedin replied:
@owenv Thank you! That would be a much better option, I'll try that out.

xedin replied:
Done!

DougGregor commented on test/Constraints/closures.swift:
```diff
@@ -349,13 +349,12 @@ takeVoidVoidFn { () -> Void in
-func f19997471(_ x: String) {}
-func f19997471(_ x: Int) {}
+func f19997471(_ x: String) {} // expected-note {{candidate expects value of type 'String' at position #0}}
+func f19997471(_ x: Int) {}    // expected-note {{candidate expects value of type 'Int' at position #0}}
```
NICE!

DougGregor commented on test/Constraints/closures.swift:
```diff
@@ -349,13 +349,12 @@ takeVoidVoidFn { () -> Void in
   takeVoidVoidFn {
-    f19997471(x) // expected-error {{cannot invoke 'f19997471' with an argument list of type '(T)'}}
-    // expected-note @-1 {{overloads for 'f19997471' exist with these partially matching parameter lists: (Int), (String)}}
+    f19997471(x) // expected-error {{no exact matches in call to global function 'f19997471'}}
```
Doesn't to be this pull request, but it would be really nice if we reminded the user of the types of the arguments here.

xedin replied:
Do you think it would be better to attach arguments to this diagnostic or to each note?

DougGregor commented on test/Constraints/diagnostics.swift:
```diff
@@ -1183,24 +1183,23 @@ func sr5081() {
 }
 
+/**
+* TODO(diagnostics):Figure out what to do when expressions are complex and completely broken
```
Would it be better to leave this enabled with FIXME'd diagnostics? That way, we'll know when we change the diagnostics, especially if we improve them.

xedin replied:
Sure, there is just going to be even more diagnostics then before but I can do that.

xedin replied:
Done!

DougGregor commented on test/Constraints/generics.swift:
```diff
@@ -161,10 +161,7 @@ class r22409190ManagedBuffer<Value, Element> {
-      // Once argument-to-parameter mismatch diagnostics are moved to the new diagnostic framework, we'll be able to restore
-      // original contextual conversion failure diagnostic here. Note that this only happens in Swift 4 mode.
-      elems.deinitialize(count: self.value)  // expected-error {{ambiguous reference to member 'deinitialize(count:)'}}
+      elems.deinitialize(count: self.value)  // expected-error {{cannot convert value of type 'UInt' to expected argument type 'Int'}}
```
Great!

DougGregor reviewed (approved):
This is fantastic!

xedin committed:
[ConstraintSystem] NFC: Fix typo `implact` -> `impact`

xedin committed:
[Diagnostics] NFC: Adjust some binary operator checks to verify exact diagnostic location

If one of the arguments to binary operator is invalid, let's
verify that diagnostic points directly to it.

xedin committed:
[Diagnostics] NFC: Restore a test-case with a TODO

xedin commented:
Please test with following pull request:
apple/swift-package-manager#2338

@swift-ci please test

xedin commented:
Ugh, have to resolve a conflict in `test/ParseableInterface/Inputs/opaque-result-types-client.swift` now...

xedin commented:
Please test with following pull request:
apple/swift-package-manager#2338

@swift-ci please test

DougGregor commented:
W00t! Merge it!

xedin commented:
Alright, let's run the tests one more time and :shipit:

xedin commented:
Please test with following pull request:
apple/swift-package-manager#2338

@swift-ci please test
